Initial state of the project

candleindark opened the pull request:
This PR include changes to set up the initial state of the project. These includes

1. Set up the project with Hatch
2. etc

candleindark committed:
Add .gitignore

candleindark committed:
Initial project setup by Hatch

candleindark committed:
Make license info for verification later

candleindark committed:
Adjust pyproject.toml from Hatch's setup

candleindark committed:
Add Hatch badge to README

candleindark committed:
Remove installation section in README

This tool is intended for internal use

candleindark committed:
Specify Python version for envs

This is current version used in developing the
translation tools. More Python versions will be
added to the types environments in the future
possibly

candleindark committed:
Set dependencies of the project

candleindark committed:
Add dandischema dependency

candleindark committed:
Remove Python spec in `types` env

The Python spec is inherited from the `default` env

candleindark committed:
Specify `dandi` as a dependency

candleindark committed:
Add CLI main entry to the project

candleindark committed:
Add CLI main entry to the project

yarikoptic commented on LICENSE.txt:
```diff
@@ -0,0 +1,9 @@
+MIT License
+
+Copyright (c) 2024-present Isaac To <candleindark@gmail.com>
```
```suggestion
Copyright (c) 2024 Isaac To <candleindark@gmail.com>
```
since eventually might/would become not correct , so best to mention real year ranges

candleindark committed:
Update LICENSE.txt

Co-authored-by: Yaroslav Halchenko <debian@onerussian.com>

candleindark committed:
Update license info per Yarik's suggestion

candleindark committed:
Correct author's email

candleindark committed:
Change the year in license info

This extends Yarik's suggestion in license info

candleindark commented on src/dandisets_linkml_status/__about__.py:
```diff
@@ -0,0 +1,4 @@
+# SPDX-FileCopyrightText: 2024 Isaac To <isaac.chun.to@gmail.com> todo: make sure this correct
+#
+# SPDX-License-Identifier: MIT  todo: make sure this correct
```
@yarikoptic Can you help input the correct license info. The current info is generated by Hatch mostly. I am perfectly fine the direct the license to DANDI. I just don't know the proper way to put it.

candleindark commented on pyproject.toml:
```diff
@@ -0,0 +1,73 @@
+description = 'A tool for reporting the status of the translation of DANDI schemas from Pydantic definitions to LinkML definitions'
+readme = "README.md"
+requires-python = ">=3.9"
+license = "MIT"  # todo: Is this the correct license?
```
@yarikoptic Can you help input the correct license info. The current info is generated by Hatch mostly. I am perfectly fine the direct the license to DANDI. I just don't know the proper way to put it.

candleindark commented on README.md:
```diff
@@ -1 +1,18 @@
+## License
+
+
+[//]: # (todo: make sure this is the correct license. Update LICENSE.TXT accordingly as well)
```
@yarikoptic Can you help input the correct license info. The current info is generated by Hatch mostly. I am perfectly fine the direct the license to DANDI. I just don't know the proper way to put it.

candleindark commented on LICENSE.txt:
```diff
@@ -0,0 +1,9 @@
+MIT License
+
+Copyright (c) 2024 Isaac To <candleindark@gmail.com>
```
@yarikoptic Can you help input the correct license info. The current info is generated by Hatch mostly. I am perfectly fine the direct the license to DANDI. I just don't know the proper way to put it.

yarikoptic replied:
Let's state

```suggestion
MIT License

Copyright (c) 2024 DANDI Team <info@dandiarchive.org>
```

candleindark committed:
Lower default env to the lowest supported Python version

candleindark committed:
Implement `tools.get_validator()`

A support function for getting the LinkML
validator of LinkML schema generated with models
in `dandischema.models`

candleindark committed:
Specify CLI options

candleindark committed:
Add `eval-type-backport` as a dependency

This is needed for interpreting some type
annotations defined in Pydantic when using
an older version of Python, e.g. 3.9.
An example of annotation is `tuple[int | str, ...]` in `pydantic_core.ErrorDetails`

candleindark committed:
Define `cli.tools.pydantic_validate()`

candleindark committed:
Define test environment

candleindark committed:
Improve docstring

candleindark committed:
Rename `cli.tools.get_validator()`

To `cli.tools.get_linkml_validator()`

candleindark committed:
Implement `cli.tools.linkml_validate()`

Function to validate dandiset metadata against
the dandiset metadata model in LinkML

candleindark committed:
Correct invocation of `pydantic2linkml.gen_linkml.translate_defs`

candleindark committed:
Define `cli.models.DandisetValidationReport`

candleindark committed:
Refactor dandiset metadata validation in LinkML

Into `cli.tools.DandisetLinkmlValidator` to
avoid repeated generation of the LinkML schema
for Dandi models

candleindark committed:
Rename model fields to shorter versions

candleindark committed:
Complete initial version of the main command

yarikoptic commented on src/dandisets_linkml_status/cli/__init__.py:
```diff
@@ -0,0 +1,73 @@
+    *,
+    include_unpublished: Annotated[
+        bool, typer.Option("--include-unpublished", "-u")
+    ] = False,
```
to not wait for querying all of them, please add an option `--dandiset` for a dandiset regex to operate just on a selection of those . We have similar one (although typically as argument, not option) in https://github.com/dandi/backups2datalad/blob/main/src/backups2datalad/__main__.py#L243 .

yarikoptic committed:
Draft for printout of results

yarikoptic commented on src/dandisets_linkml_status/cli/__init__.py:
```diff
@@ -0,0 +1,79 @@
+                )
+            )
+    # import pdb; pdb.set_trace()
+    print('\n'.join(f"dandiset: {r.dandiset_identifier}, linkml: {len(r.linkml_validation_errs or [])}, pydantic: {len(r.pydantic_validation_errs or [])}" for r in validation_reports))
```
refactor this so it prints into e.g. STATUS.md file as markdown table

Also add a column with dandiset version.

All counts should lead to the corresponding file `{dandiset_id}/linkml-errors.yaml` or `pydanditc-errors.yaml` there pretty printed.

candleindark replied:
Done.

candleindark committed:
Remove unneeded commented code

candleindark committed:
Remove unneeded comment

candleindark committed:
Fix incorrect assignment of `linkml_validation_errs` to `None`

candleindark committed:
Temporarily change pydantic2linkml dependency to a different branch

For testing during development

candleindark committed:
Introduce `--output-file` option to CLI

candleindark committed:
Write validation reports to the output file

candleindark committed:
Improve help text for the `--output-file` option

candleindark committed:
Define a smoke test of the CLI

candleindark committed:
Improve `DandisetValidationReport`

There is no need to use Optional types in the
error fields. An empty list is sufficient to
express nonexistence of errors

candleindark committed:
Add a todo for future debugging

candleindark committed:
Switch the pydantic2linkml dependency back to the enh branch

candleindark committed:
Define the interface of `output_reports`

candleindark committed:
Remove unneeded import

candleindark committed:
Implement output of reports to directories

candleindark committed:
Rename output directory for validation reports

candleindark committed:
RF: Refactor code for writing validation results to files

candleindark committed:
RF: Create needed `TypeAdapter` in models.py

Creation of `TypeAdapter` has non-trivial performance impact.
This change eliminate repeating creation of the same
`TypeAdapter` objects

candleindark committed:
Remove writing of reports to validation_reports.json

Reports are now stored in the `reports` directory

candleindark committed:
Add `pyyaml` dependency

Needed for outputting results in yaml format

candleindark committed:
Output validation report in yaml format as well

candleindark committed:
Add `--dandi-instance` option to CLI

This also decides the output dir based on the given
dandi instance

candleindark committed:
Output report summary to `summary.md` instead of `README.md`

candleindark committed:
Write table header to report summary

candleindark committed:
Change the output put directory name

candleindark committed:
RF: Implement function to generate a row in summary table

candleindark committed:
Populate the dandiset column in the summary table

candleindark committed:
Add dandiset version to validation report

candleindark committed:
Reorganise validation reports according to dandiset version

candleindark committed:
Populate the version column of the summary table

candleindark committed:
RF: Introduce the `row_cells` variable

This can eliminate an indent and the variable name also serves as
a documentation

candleindark committed:
Include `schema_version` as computed field in `DandisetValidationReport`

candleindark committed:
Populate schema version column in summary table

candleindark committed:
Rename `schema_version` to `dandiset_schema_version` in `DandisetValidationReport`

candleindark committed:
Add dandiset version status and modified datetime to `DandisetValidationReport`

candleindark committed:
Populate `api_status` and `modified` column in summary table

candleindark committed:
Unset computed field status of `dandiset_schema_version`

candleindark committed:
Populate the `pydantic` column of the summary table

candleindark committed:
RF: Introduce `dandiset_dir` var

To make code more readable, such as depicting
the relationship between different paths

candleindark committed:
Remove link for Pydantic column when no error

candleindark committed:
Introduce and use `isorted()`

candleindark committed:
RF: Introduce and use `get_pydantic_err_counts()`

candleindark committed:
Outline `get_linkml_err_counts()`

candleindark committed:
Rename `DANDISET_METADATA_TYPE` to `DandisetMetadataType`

candleindark committed:
Rename `PYDANTIC_VALIDATION_ERRS_TYPE` to `PydanticValidationErrsType`

candleindark committed:
Rename `LINKML_VALIDATION_ERRS_TYPE` to `LinkmlValidationErrsType`

candleindark committed:
Rename `DANDISET_METADATA_ADAPTER` to `dandiset_metadata_adapter`

candleindark committed:
Rename `PYDANTIC_VALIDATION_ERRS_ADAPTER` to `pydantic_validation_errs_adapter`

candleindark committed:
Rename `LINKML_VALIDATION_ERRS_ADAPTER` to `linkml_validation_errs_adapter`

candleindark committed:
RF: Move `gen_row()` to module level as `_gen_row()`

Defining the function as an inner function makes
the point of interaction of the function less obvious

candleindark committed:
Remove unneeded comments

candleindark committed:
RF: Move `write_data()` inner func to module level

This makes the interaction of the func more obvious

candleindark committed:
Complete implementation of `get_linkml_err_counts()`

candleindark committed:
Populate the `linkml` column in summary table

candleindark committed:
Add spaces around the row cell contents

For better readability

candleindark committed:
Log completion of output of validation reports

candleindark committed:
RF: Refactor code into `compile_validation_report()`

candleindark committed:
Output reports for both draft and latest versions on published dandisets

candleindark committed:
Add `__pycache__` to `.gitignore`

yarikoptic commented:
wow -- it turned out to be truly huge, forbidding review of the code... can only go by commits. May be it would be vise to dump them into a separate repo (e.g. `https://github.com/dandi/dandisets-linkml-status-reports`) which would have `code/` folder pointing to this as git submodule or any other way you like, so please remove that last commit (`git reset --hard HEAD^; git push -f`)

Otherwise -- let's aim to merge to finalize the PR so we could then review the errors and target to address some of linkml gotchas (odd that it detects more errors than pydantic)

yarikoptic commented:
also I think we better do not dump full instance record into error files like https://github.com/candleindark/dandisets-linkml-status/blob/init/dandi-reports/000003/0.230629.1955/linkml_validation_errs.yaml -- makes it impossible to see actual errors.  We have metadata file nearby handy if needed: https://github.com/candleindark/dandisets-linkml-status/blob/init/dandi-reports/000003/0.230629.1955/metadata.yaml

candleindark committed:
Make outputting of validation reports on draft version unconditional

For rational, https://github.com/dandi/dandisets-linkml-status/issues/2#issuecomment-2347086807

candleindark commented:
> wow -- it turned out to be truly huge, forbidding review of the code... can only go by commits. May be it would be vise to dump them into a separate repo (e.g. `https://github.com/dandi/dandisets-linkml-status-reports`) which would have `code/` folder pointing to this as git submodule or any other way you like, so please remove that last commit (`git reset --hard HEAD^; git push -f`)

FYI: The validation reports are now removed from the repo.
